### PR TITLE
Equal branches: master <- develop

### PR DESCRIPTION
**Pull request by opalczynski**

### CODE REVIEW

File: api_sketch/constants.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    'positiveinteger': 'PositiveIntegerField',
+    'positivesmallinteger': 'PositiveSmallIntegerField',
+    'slug': 'SlugField',
+    'smallinteger': 'SmallIntegerField',
+    'text': 'TextField',
+    'time': 'TimeField',
+    'url': 'URLField',
+    'uuid': 'UUIDField',
+    'foreign_key': 'ForeignKey',
+    'manytomany': 'ManyToManyField',
```
Comment (opalczynski) [inline api_sketch/constants.py]: Maybe we should stick to the camel case?

File: api_sketch/constants.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    'onetomany': 'OneToOneField',
+}
+
+CRUD_METHODS = {
+    'create', 'read', 'update', 'delete'
```
Comment (opalczynski) [inline api_sketch/constants.py]: add list here also

File: api_sketch/constants.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+GENERIC_FIELD_OPTIONS = {
+    'null', 'blank', 'choices', 'db_column', 'db_index', 'db_tablespace',
+    'default', 'editable', 'error_messages', 'help_text', 'primary_key',
+    'unique', 'unique_for_date', 'unique_for_month', 'unique_for_year',
```
Comment (opalczynski) [inline api_sketch/constants.py]: isn't the `unique_for_date`, `unique_for_month`, `unique_for_year` DateTime specific?

File: api_sketch/constants.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    'GenericIPAddressField': {'protocol', 'unpack_ipv4'},
+    'SlugField': {'max_length'},
+    'TimeField': {'auto_now', 'auto_now_add'},
+    'URLField': {'max_length'},
+    'ForeignKey': {'othermodel', 'on_delete'},
+    'ManyToManyField': {'othermodel'},
+    'OneToOneField': {'othermodel', 'on_delete', 'parent_link'},
+}
+
+def get_allowed_options(field_type):
```
Comment (opalczynski) [inline api_sketch/constants.py]: This is not constant - so shouldn't be in the constant file.

File: api_sketch/constants.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    return GENERIC_FIELD_OPTIONS | FIELD_SPECIFIC_OPTIONS.get(field_type, set())
+
+def get_all_field_names():
```
Comment (opalczynski) [inline api_sketch/constants.py]: as above

File: api_sketch/constants.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    return {type_name for (slug, type_name) in list(FIELD_TYPES.items())}
+
+def get_all_api_methods():
```
Comment (opalczynski) [inline api_sketch/constants.py]: And this can be Simple Set I think - if you add `list` to CRUD

File: api_sketch/model.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        self.api_methods = api_methods
+        self.validate()
+
+    def validate(self):
+        for api_method in self.api_methods:
+            if api_method not in get_all_api_methods():
+                raise ValidationError
+
+
+def model_from_dict(name, data):  # TODO exception handling looks awful
```
Comment (opalczynski) [inline api_sketch/model.py]: agree :)

File: tests/testparser.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  api:
+    - read
+'''
+
+class ParseInvalidModels(unittest.TestCase):
+    def setUp(self):
+        self.processor = Processor()
+
+    def test_missing_name(self):
+        self.processor = Processor()
```
Comment (opalczynski) [inline tests/testparser.py]: Why are you creating processor in each test - if you have it in setUp?

File: api_sketch/constants.py (hunk 4 above)
Comment (Glenpl) [inline api_sketch/constants.py]: Actually it's a constant expression

**Comment by opalczynski** [comment]
Looks promising :)

File: tests/testparser.py (hunk 8 above)
Comment (Glenpl) [inline tests/testparser.py]: Yeah, I forgot to remove this and I was copying & pasting these tests - I've already fixed it locally :)

File: api_sketch/constants.py (hunk 3 above)
Comment (Glenpl) [inline api_sketch/constants.py]: https://docs.djangoproject.com/en/1.9/ref/models/fields/#field-options

no, according to docs they aren't

File: api_sketch/constants.py (hunk 2 above)
Comment (Glenpl) [inline api_sketch/constants.py]: https://en.wikipedia.org/wiki/Create,_read,_update_and_delete

>The acronym may be extended to CRUDL to cover listing

should I rename it to `CRUDL` then?

File: api_sketch/constants.py (hunk 1 above)
Comment (Glenpl) [inline api_sketch/constants.py]: like `manyToMany`?

Comment (Glenpl) [inline api_sketch/constants.py]: ok, I get it, actually it's meaningless -  in the YAML file you can write it even as `MaNyToMaNy` - these names get lowered somewhere during processing :)

Comment (opalczynski) [inline api_sketch/constants.py]: Shiet :) I'm tired. Thinking of snake_case :P writing camelCase :)

foreign_key - is nice, snake
manytomany - many_to_many will be nicer :) I think. But this is not so important at current level of dev process.

File: api_sketch/constants.py [hunk 9]
```diff
@@ -27,13 +25,13 @@
     'time': 'TimeField',
     'url': 'URLField',
     'uuid': 'UUIDField',
-    'foreign_key': 'ForeignKey',
+    'foreignkey': 'ForeignKey',
```
Comment (lukaszjurczak) [inline api_sketch/constants.py]: This one was ok. Just change the m2m and one2m to snake_case as opalczynski said. But yeah, it's so important for now :P

File: api_sketch/constants.py [hunk 10]
```diff
@@ -27,13 +25,13 @@
...
     'manytomany': 'ManyToManyField',
     'onetomany': 'OneToOneField',
```
Comment (lukaszjurczak) [inline api_sketch/constants.py]: You don't need any one_to_many. It's just foreign_key.

File: api_sketch/constants.py (hunk 9 above)
Comment (Glenpl) [inline api_sketch/constants.py]: He said we should stick to camel case; as I said before, in the YAML file you can capitalize whichever letters you want - processing & validation is case insensitive :)

Comment (lukaszjurczak) [inline api_sketch/constants.py]: He said that it was his mistake and you _should_ stick to snake :P

File: api_sketch/constants.py (hunk 10 above)
Comment (Glenpl) [inline api_sketch/constants.py]: Yeah, it's my mistake - this line should look like this: `'onetoone': 'OneToOneField',` - thanks for pointing this out for me!

File: api_sketch/constants.py (hunk 9 above)
Comment (Glenpl) [inline api_sketch/constants.py]: I haven't noticed he'd added another comment - a point for you :P

File: api_sketch/settings.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        self['debug'] = True
+        self['root_urlconf'] = project_name + '.urls'
+        self['allowed_hosts'] = set()
+        self['wsgi_application'] = project_name + '.wsgi.application'
+        self['internationalization'] = DEFAULT_INTERNATIONALIZATION
+        self['drf_permission_classes'] = DEFAULT_DRF_PERMISSION_CLASSES
+        self['static_url'] = r'/static/'
+
+    # TODO extract validating functions
+    def add(self, name, data):
```
Comment (opalczynski) [inline api_sketch/settings.py]: :) Nice if else block.

But try to rewrite it in this way:

```
def add(self, name, data):
    self[name] = handlers.get(name)(data)

# Where handelrs will be such structure:

handlers = {
    'database': database_handler
    # ...
}

def database_handler(data):
      if data.get('engine') not in DATABASE_ENGINES:
           raise ValueError('unknown database engine')
      data['engine'] = DATABASE_ENGINES[data['engine']]
      return data    

# or something similar :) Now it looks like not manageable one;
# you can also define a separate class for such handlers
```

File: resources/model.jinja2 [hunk 12]
```diff
@@ -0,0 +1,9 @@
+class {{ model.name }}(models.Model):
```
Comment (opalczynski) [inline resources/model.jinja2]: ++ for using jinja here!

File: tests/testparser.py [hunk 13]
```diff
@@ -0,0 +1,358 @@
+import unittest
+
+from api_sketch.processor import Processor
+
+valid_single_model = '''
```
Comment (opalczynski) [inline tests/testparser.py]: Define all this in seperate file - and then assing to self in setUp :) Will be cleaner. Let say that: tests/yml_file_templates.py

**Comment by opalczynski** [comment]
I will merge it - but take a look at the comments :)

**Pull request merged**